prov/efa/src/efa_hmem.c

```diff
@@ -377,11 +337,14 @@ int efa_domain_hmem_validate_p2p_opt(struct efa_domain *efa_domain, enum fi_hmem
 	 *     PREFERED means a provider should prefer P2P if it is available.
 	 *
 	 * These options does not require that p2p is supported by device,
-	 * nor do they prohibit that p2p is reqruied by implementation. Therefore
-	 * they are always supported.
+	 * nor do they prohibit that p2p is required by implementation. 
+	 * Therefore they are always supported unless p2p is disabled.
 	 */
 	case FI_HMEM_P2P_PREFERRED:
 	case FI_HMEM_P2P_ENABLED:
+		if (OFI_UNLIKELY(ofi_hmem_p2p_disabled()))
+			return -FI_EOPNOTSUPP;
+
 		info->p2p_disabled_by_user = false;
```

Comment:
I don't think we need to handle such case regardless of info->p2p_disabled_by_peer.

>FI_HMEM_P2P_ENABLED: Peer to peer support may be used by the provider to handle FI_HMEM transfers, and which transfers are initiated using peer to peer is subject to the provider implementation.
FI_HMEM_P2P_PREFERRED: Peer to peer support should be used by the provider for all transfers if available, but the provider may choose to copy the data to initiate the transfer if peer to peer support is unavailable.


Based on the definition, the provider can decide which memcpy protocol should be used. In other words, if p2p is disabled by the user, EFA provider can decide to use a different protocol.

Also, we shouldn't `info->p2p_disabled_by_user = false;` here, as @darrylabbate already mentioned, the `FI_HMEM_DISABLE_P2P` option has a higher priority. It's much easier to change an environment variable than fi_setopt code.